toolkit/smart-contracts/offchain/src/key_value/mod.rs

```diff
@@ -0,0 +1,185 @@
+use crate::csl::{
+	empty_asset_name, get_builder_config, unit_plutus_data, CostStore, Costs,
+	TransactionBuilderExt, TransactionExt,
+};
+use crate::governance::GovernanceData;
+use crate::multisig::submit_or_create_tx_to_sign;
+use crate::multisig::MultiSigSmartContractResult::TransactionSubmitted;
+use crate::plutus_script::PlutusScript;
+use crate::{
+	await_tx::{AwaitTx, FixedDelayRetries},
+	cardano_keys::CardanoPaymentSigningKey,
+	csl::TransactionContext,
+	multisig::MultiSigSmartContractResult,
+};
+use anyhow::anyhow;
+use cardano_serialization_lib::{PlutusData, Transaction, TransactionBuilder};
+use ogmios_client::{
+	query_ledger_state::{QueryLedgerState, QueryUtxoByUtxoId},
+	query_network::QueryNetwork,
+	transactions::Transactions,
+	types::OgmiosUtxo,
+};
+use partner_chains_plutus_data::key_value::{key_value_to_plutus_data, KeyValueDatum};
+use sidechain_domain::byte_string::ByteString;
+use sidechain_domain::{KeyValue, PolicyId, UtxoId};
+
+pub trait KeyValueStore {
+	#[allow(async_fn_in_trait)]
+	async fn insert_key_value(
+		&self,
+		genesis_utxo: UtxoId,
+		key: String,
+		value: ByteString,
+		payment_signing_key: &CardanoPaymentSigningKey,
+	) -> anyhow::Result<Option<MultiSigSmartContractResult>>;
+}
+
+impl<C: QueryLedgerState + QueryNetwork + Transactions + QueryUtxoByUtxoId> KeyValueStore for C {
+	async fn insert_key_value(
+		&self,
+		genesis_utxo: UtxoId,
+		key: String,
+		value: ByteString,
+		payment_signing_key: &CardanoPaymentSigningKey,
+	) -> anyhow::Result<Option<MultiSigSmartContractResult>> {
+		insert_key_value(
+			genesis_utxo,
+			key,
+			value,
+			payment_signing_key,
+			self,
+			&FixedDelayRetries::two_minutes(),
+		)
+		.await
+	}
+}
+
+pub async fn insert_key_value<
+	C: QueryLedgerState + QueryNetwork + Transactions + QueryUtxoByUtxoId,
+	A: AwaitTx,
+>(
+	genesis_utxo: UtxoId,
+	key: String,
+	value: ByteString,
+	payment_signing_key: &CardanoPaymentSigningKey,
+	ogmios_client: &C,
+	await_tx: &A,
+) -> anyhow::Result<Option<MultiSigSmartContractResult>> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key, ogmios_client).await?;
+	let (validator, policy) =
+		crate::scripts_data::generic_container_scripts(genesis_utxo, ctx.network)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let validator_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+
+	let tx_hash_opt = match get_current_key_value(validator_utxos, key.clone(), policy.policy_id())?
+	{
+		Some(current_value) => {
+			log::info!("There is already a value stored for key '{}': {:?}", key, current_value);
+			None
```

Comment:
My take on Insert API is that it can be idempotent:
`current_value == value => Ok(None)`,
but `current_value == value => Err("Another value is at this key already")`